### PR TITLE
Replace history state when doing auto query correction

### PR DESCRIPTION
**Pull request by olamothe**
Replace history state will not create a new history entry, and so it will be possible to "pop" from the last browser state properly without getting stuck in a loop where you are going back to an invalid query that return corrections automatically.

https://coveord.atlassian.net/browse/JSUI-2058

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: unitTests/ui/DidYouMeanTest.ts [hunk 1]
```diff
@@ -65,61 +65,89 @@ export function DidYouMeanTest() {
...
-          test.env.queryStateModel.get = () => 'foobar';
-          var qsmSpy = jasmine.createSpy('queryStateModelSpy');
-          test.env.queryStateModel.set = qsmSpy;
 
-          Simulate.query(test.env, {
-            results: FakeResults.createFakeResults(0),
-            queryCorrections: [fakeQueryCorrection]
-          });
+          it('should autocorrect the query when no results are found', () => {
+            const qsmSpy = jasmine.createSpy('queryStateModelSpy');
```
Comment (samisayegh) [inline unitTests/ui/DidYouMeanTest.ts]: I think it is better to not abbreviate the variable name.

File: unitTests/ui/DidYouMeanTest.ts [hunk 2]
```diff
@@ -65,61 +65,89 @@ export function DidYouMeanTest() {
...
+            const analyticsSpy = jasmine.createSpy('analyticsSpy');
+            test.cmp.usageAnalytics.logSearchEvent = analyticsSpy;
+
+            Simulate.query(test.env, {
+              results: FakeResults.createFakeResults(0),
+              queryCorrections: [fakeQueryCorrection]
+            });
+
+            expect(analyticsSpy).toHaveBeenCalledWith(analyticsActionCauseList.didyoumeanAutomatic, {});
+            expect(analyticsSpy.calls.count()).toBe(1);
```
Comment (samisayegh) [inline unitTests/ui/DidYouMeanTest.ts]: Would it be possible to use the method `toHaveBeenCalledTimes` for this? e.g.

```
expect(analyticsSpy).toHaveBeenCalledTimes(1)
```

File: unitTests/ui/DidYouMeanTest.ts [hunk 3]
```diff
@@ -65,61 +65,89 @@ export function DidYouMeanTest() {
...
-          expect(analyticsSpy.calls.count()).toBe(1);
+          it('should replace the state in history manager when query is autocorrected', () => {
+            spyOn(test.env.searchInterface.historyManager, 'replaceState');
+
+            Simulate.query(test.env, {
+              results: FakeResults.createFakeResults(0),
+              queryCorrections: [fakeQueryCorrection]
+            });
+
+            expect(test.env.searchInterface.historyManager.replaceState).toHaveBeenCalled();
```
Comment (samisayegh) [inline unitTests/ui/DidYouMeanTest.ts]: Small point, but I think it is better to use `toHaveBeenCalledTimes` for void methods as a more strict check than `toHaveBeenCalled` to make sure no unexpected calls are being made.

File: unitTests/ui/DidYouMeanTest.ts [hunk 4]
```diff
@@ -65,61 +65,89 @@ export function DidYouMeanTest() {
...
+            expect(test.env.searchInterface.historyManager.replaceState).not.toHaveBeenCalled();
+          });
         });
       });
     });
 
-    it('should not autocorrect search-as-you-type queries', function() {
-      var qsmSpy = jasmine.createSpy('queryStateModelSpy');
+    it('should not autocorrect search-as-you-type queries', () => {
+      const qsmSpy = jasmine.createSpy('queryStateModelSpy');
```
Comment (samisayegh) [inline unitTests/ui/DidYouMeanTest.ts]: Same as above, my preference would be to have better readability than to abbreviate

File: src/controllers/HistoryManager.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+export interface IHistoryManager {
+  setState(state: Record<string, any>): void;
```
[the remark this answers is not in the document]
Comment (olamothe) [reply src/controllers/HistoryManager.ts]: Yeah, the `return` is returning void anyway, so it's kind of useless, true.

File: src/controllers/HistoryManager.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+  replaceState(state: Record<string, any>): void;
```
[the remark this answers is not in the document]
Comment (olamothe) [reply src/controllers/HistoryManager.ts]: Idem, I'll replace.

File: unitTests/ui/DidYouMeanTest.ts (hunk 1 above)
Comment (olamothe) [reply unitTests/ui/DidYouMeanTest.ts]: Good point, I was merely copy pasting the old tests, but I can take the time to rename them.

**Pull request merged**